src/app/clusters/device-energy-management-server/device-energy-management-server.cpp

```diff
@@ -665,15 +847,47 @@ void Instance::HandleRequestConstraintBasedForecast(HandlerContext & ctx,
 
 void Instance::HandleCancelRequest(HandlerContext & ctx, const Commands::CancelRequest::DecodableType & commandData)
 {
-    Status status;
+    Status status                                               = Status::Failure;
+    DataModel::Nullable<Structs::ForecastStruct::Type> forecast = mDelegate.GetForecast();
+
+    if (forecast.IsNull())
+    {
+        ChipLogDetail(AppServer, "Cancelling on a Null forecast!");
+        status = Status::Failure;
+    }
+    else if (forecast.Value().forecastUpdateReason == ForecastUpdateReasonEnum::kInternalOptimization)
+    {
+        ChipLogDetail(AppServer, "Bad Cancel when ESA ForecastUpdateReason was already Internal Optimization!");
+        status = Status::InvalidInState;
+    }
+    else
+    {
+        status = mDelegate.CancelRequest();
+    }
 
-    status = mDelegate.CancelRequest();
     ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
-    if (status != Status::Success)
+}
+
+Status Instance::GetCurrentUtcTime(uint32_t & currentUtcTime) const
+{
+    currentUtcTime = 0;
+    System::Clock::Milliseconds64 cTMs;
+
+    CHIP_ERROR err = System::SystemClock().GetClock_RealTimeMS(cTMs);
+    if (err != CHIP_NO_ERROR)
     {
-        ChipLogError(Zcl, "DEM: CancelRequest FAILURE");
-        return;
+        ChipLogError(Zcl, "DEM: Unable to get current time - err:%" CHIP_ERROR_FORMAT, err.Format());
+        return Status::Failure;
+    }
+
+    auto unixEpoch = std::chrono::duration_cast<System::Clock::Seconds32>(cTMs).count();
+    if (!UnixEpochToChipEpochTime(unixEpoch, currentUtcTime))
```

Comment:
It's very much not obvious based on naming that this function returns Matter epoch times.  Please rename accordingly.

Reply:
I believe UnixEpochToChipEpochTime is defined in connectedhomeip/src/lib/support/TimeUtils.h and used in a few places in connectedhomeip (see below) -> do you still want me to rename it?

connectedhomeip/examples/chef/common/clusters/door-lock/chef-lock-endpoint.cpp
connectedhomeip/examples/lock-app/lock-common/src/LockEndpoint.cpp
connectedhomeip/src/credentials/CHIPCertificateSet.h
connectedhomeip/src/tools/chip-cert/Cmd_ValidateCert.cpp

If I do rename it, would GetMatterEpochTimeFromUnixTime() be ok?

Reply:
UnixEpochToChipEpochTime is fine as far as it goes.  I meant the function here, which calls that and returns a Matter epoch time, but does not make that clear.